File: src/SampleCRM.Web/Services/OrderItemsService.cs
using OpenRiaServices.DomainServices.Hosting;
using OpenRiaServices.DomainServices.Server;
using SampleCRM.Web.Attributes;
using SampleCRM.Web.Models;
using System;
using System.Data.Entity.Migrations;
using System.Linq;

namespace SampleCRM.Web
{
    [EnableClientAccess]
    public class OrderItemsService : SampleCRMService
    {
        [Query]
        public IQueryable<OrderItem> GetOrderItems(string search) => 
            _context.OrderItems.OrderBy(o => o.OrderLine)
                           .Where(x => x.ProductID.ToString().Contains(search) 
                                       || x.OrderID.ToString().Contains(search)
                                       || search == "")
                           .AsQueryable();

        public OrderItem GetOrderItemById(int orderId, int orderLine) => 
            GetOrderItems(string.Empty)
                .SingleOrDefault(x => x.OrderID == orderId && x.OrderLine == orderLine);

        [Query]
        public IQueryable<OrderItem> GetOrderItemsOfOrder(long orderId, string search) => 
            GetOrderItems(search)
                .Where(x => x.OrderID == orderId);

        [Query]
        public IQueryable<OrderItem> GetOrderItemsOfProduct(string productId, string search) => 
            GetOrderItems(search)
                .Where(x => x.ProductID == productId);

        [Delete]
        [RestrictAccessReadonlyMode]
        public void DeleteOrderItem(OrderItem orderItem)
        {
            var dOrderItem = _context.OrderItems
                                     .FirstOrDefault(x => x.OrderID == orderItem.OrderID && x.OrderLine == orderItem.OrderLine);
            if (dOrderItem == null)
                return;

            _context.OrderItems.Remove(dOrderItem);
            _context.SaveChanges();
        }

        [Insert]
        [RestrictAccessReadonlyMode]
        public void InsertOrderItem(OrderItem orderItem)
        {
            var lastItem = _context.OrderItems
                                   .Where(x => x.OrderID == orderItem.OrderID)
                                   .OrderByDescending(x => x.OrderLine)
                                   .FirstOrDefault();
            if (lastItem == null)
            {
                orderItem.OrderLine = 1;
            }
            else
            {
                orderItem.OrderLine = lastItem.OrderLine + 1;
            }

            _context.OrderItems.Add(orderItem);

#if DEBUG
            var validationResult = _context.Entry(orderItem).GetValidationResult();
            if (validationResult.ValidationErrors.Any())
            {
                Console.WriteLine($"Validation Error in InsertOrderItem: {validationResult.ValidationErrors.FirstOrDefault().PropertyName} {validationResult.ValidationErrors.FirstOrDefault().ErrorMessage}");
            }
#endif

            _context.SaveChanges();
        }

        [Update]
        [RestrictAccessReadonlyMode]
        public void UpdateOrderItem(OrderItem orderItem)
        {
            _context.OrderItems.AddOrUpdate(orderItem);
            _context.SaveChanges();
        }
    }
}

File: src/SampleCRM.Web/Services/OrderItemsService.vb
Imports OpenRiaServices.DomainServices.Hosting
Imports OpenRiaServices.DomainServices.Server
Imports SampleCRM.Web.Attributes
Imports System.Data.Entity.Migrations
Imports System.Linq

Namespace SampleCRM.Web
    <EnableClientAccess>
    Public Class OrderItemsService
        Inherits SampleCRMService

        <Query>
        Public Function GetOrderItems() As IQueryable(Of OrderItems)
            Return _context.OrderItems
        End Function

        Public Function GetOrderItemById(ByVal orderId As Integer) As OrderItems
            Return _context.OrderItems.SingleOrDefault(Function(x) x.OrderID = orderId)
        End Function

        <Query>
        Public Function GetOrderItemsOfOrder(ByVal orderId As Long) As IQueryable(Of OrderItems)
            Dim itemsQuery = _context.OrderItems.Where(Function(x) x.OrderID = orderId)
            Return itemsQuery
        End Function

        <Query>
        Public Function GetOrderItemsOfProduct(ByVal productId As String) As IQueryable(Of OrderItems)
            Return _context.OrderItems.Where(Function(x) x.ProductID = productId)
        End Function

        <Delete>
        <RestrictAccessDeveloperMode>
        Public Sub DeleteOrderItem(ByVal orderItem As OrderItems)
            Dim dOrderItem = _context.OrderItems.FirstOrDefault(Function(x) x.OrderID = orderItem.OrderID AndAlso x.OrderLine = orderItem.OrderLine)
            If dOrderItem Is Nothing Then Return
            _context.OrderItems.Remove(dOrderItem)
            _context.SaveChanges()
        End Sub

        <Insert>
        <RestrictAccessDeveloperMode>
        Public Sub InsertOrderItem(ByVal orderItem As OrderItems)
            Dim lastItem = _context.OrderItems.Where(Function(x) x.OrderID = orderItem.OrderID).OrderByDescending(Function(x) x.OrderLine).FirstOrDefault()

            If lastItem Is Nothing Then
                orderItem.OrderLine = 1
            Else
                orderItem.OrderLine = lastItem.OrderLine + 1
            End If

            _context.OrderItems.Add(orderItem)
            _context.SaveChanges()
        End Sub

        <Update>
        <RestrictAccessDeveloperMode>
        Public Sub UpdateOrderItem(ByVal orderItem As OrderItems)
            _context.OrderItems.AddOrUpdate(orderItem)
            _context.SaveChanges()
        End Sub
    End Class
End Namespace
